File: Python/utils/searching.py
import collections


# Linear Search
# Iterative
def linear_search(list, key):
    for i in range(len(list)):
        if list[i] == key:
            return i
    return -1


# Recursive
def linear_search_rec(list, key, size):
    if size == 0:
        return -1
    elif list[size - 1] == key:
        return size - 1
    else:
        return linear_search_rec(list, key, size - 1)


# Binary Search
# Iterative implementation
def binary_search_iterative(v, x):
    l = 0
    h = len(v) - 1
    while h - l > 1:
        mid = (h + l) // 2
        if v[mid] < x:
            l = mid + 1
        else:
            h = mid
    if v[l] == x:
        print(f'Element found at index {l}')
    elif v[h] == x:
        print(f'Element found at index {h}')
    else:
        print('Element not found')


# Recursive implementation
def binary_search_rec(arr, l, r, x):
    if r >= l:
        mid = l + (r - 1) // 2
        if arr[mid] == x:
            return mid
        elif arr[mid] > x:
            return binary_search_rec(arr, l, mid-1, x)
        else:
            return binary_search_rec(arr, mid + 1, r, x)
    else:
        return -1


# Breadth First Search
def bfs(graph, root):
    visited, queue = set(), collections.deque([root])
    visited.add(root)
    while queue:
        vertex = queue.popleft()
        print(str(vertex) + " ", end="")
        for neighbour in graph[vertex]:
            if neighbour not in visited:
                visited.add(neighbour)
                queue.append(neighbour)


# Depth First Search
def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start)
    for next in graph[start] - visited:
        dfs(graph, next, visited)
    return visited

File: rust/src/utils/searching.rs
/* Searching Algorithms */
use std::collections::LinkedList;

// Linear Search
pub fn linear_search(arr: &[i32], key: &i32) -> Option<usize> {
    for (i, data) in arr.iter().enumerate() {
        if key == data {
            return Some(i);
        }
    }
    None
}

// Binary Search
pub fn binary_search(arr: &[i32], find: i32) -> Option<usize> {
    let length = arr.len();
    let mut half_index = length / 2;
    let mut high_index = length - 1;
    let mut low_index = 0;
    let mut current = arr[half_index];
    while low_index <= high_index {
        match current.cmp(&find) {
            std::cmp::Ordering::Equal => return Some(half_index),
            std::cmp::Ordering::Less => low_index = half_index +1,
            std::cmp::Ordering::Greater => high_index = half_index -1,
        }
        half_index = (high_index + low_index) / 2;
        current = arr[half_index];
    }
    return None;
}


// Graph structure
pub struct Graph {
    pub vertices: Vec<Vec<u32>>,
}

impl Graph {
    pub fn new(size: usize) -> Self {
        Graph { vertices: vec![Vec::new(); size] }
    }

    pub fn add_edge(&mut self, u: usize, v: usize) {
        self.vertices[u].push(v as u32);
    }
}

// BFS implementation
pub fn bfs(graph: &Graph, start: usize) {
    let mut visited = vec![false; graph.vertices.len()];
    let mut queue = LinkedList::new();

    visited[start] = true;
    queue.push_back(start);

    while !queue.is_empty() {
        let u = queue.pop_front().unwrap();
        print!("{} ", u);

        for v in &graph.vertices[u] {
            if !visited[*v as usize] {
                visited[*v as usize] = true;
                queue.push_back(*v as usize);
            }
        }
    }
}

// DFS implementation
pub fn dfs(graph: &Graph, start: usize, visited: &mut Vec<bool>) {
    visited[start] = true;
    print!("{} ", start);

    for v in &graph.vertices[start] {
        if !visited[*v as usize] {
            dfs(graph, *v as usize, visited);
        }
    }
}

// BFS Breadth First Search


// DFS Depth First Search
